### ISSUE TITLE
Where clauses

### ISSUE THREAD

**Issue by nrc** [rust-lang-nursery/fmt-rfcs#38]
In functions, structs, enums, traits, impls, etc.

**Comment by nrc** [comment]
Several comments on where clauses, starting as https://github.com/rust-lang-nursery/fmt-rfcs/issues/30#issuecomment-258113982

**Comment by joshtriplett** [comment]
One note that came up about `where` regarding #8 in the rust-style meeting: some of the arguments for block indentation rather than visual indentation apply less to visual indentation by the width of a keyword.  For instance, in the following:

```rust
fn f<T, U, V>{...)
    where T: ...,
          U: ...,
          V: ...
{
    ...
}

```

The visual alignment of `T`, `U`, and `V` causes less of a problem than other forms of visual indent, since the width of `where` won't change.

This is not indented as an argument for/against block/visual indentation, just an observation.

**Comment by solson** [comment]
> This is not indented as an argument

I see what you did there.

**Comment by withoutboats** [comment]
Just wanted to inform y'all that I'm amending Rust RFC #1598 (associated type constructors) to include a syntax for higher rank types in which you attach a turbofish to the `where` keyword. If this is accepted I imagine it will impact the formatting choice you make here.

An example of the new syntax:

```rust
fn foo<F>() where::<'a> F: Fn(&'a T) { }
```

**Comment by nrc** [comment]
I'd like to propose (somewhat roughly) the following:
* a where clause should always start a new line and be block indented one level deeper than its 'parent'
* if there is more than one sub-clause, then put each sub-clause on its own line. Sub-clauses should be visually indented.
* there is no trailing comma (I'm not sure about this, I've just never done it myself, perhaps there should be one)
* when a where clause is followed by a block, the opening brace is put on a new line and is block indented to the same depth as the where clause's parent
* within sub-clauses, format as bounds elsewhere (is this all defined in the guide alread?), in particular, spaces around `=` and `+`, and only after `:`.

The example in https://github.com/rust-lang-nursery/fmt-rfcs/issues/38#issuecomment-260505823 satisfies these guidelines.

**Comment by solson** [comment]
I would include the trailing comma. You made a statement on another issue (I forget which) about including all trailing commas that are followed by newlines, and I would apply the same rule here.

**Comment by joshtriplett** [comment]
I would include the trailing comma as well, for the multi-line case, though not in the single-line case (even though it immediately precedes a newline, since where clauses have no surrounding braces).

I do agree with putting a newline before the opening `{` when you have a where clause.

I'll mildly argue against visual indent here, but only mildly, given the fixed-width observation I made in #38.

I'd assume, from your mention of visual indent, that when you say "each sub-clause on its own line", the first sub-clause shares the same line as the `where`?

One other observation: in the case where a function has no return value, I rather like this style:

```rust
fn function_name<T, U>(
    thing: T,
    uthing: U,
) where ...
```

However, that style does not naturally extend to the return-value case, where I'd prefer a newline before `where`.  So, for consistency, I agree that we should always require a newline before `where`.

Finally, I'd suggest explicitly saying "a where clause should always start a new line, *even when it would fit on the previous line*".  I don't care for `fn shortfunc<T, U>(thing: T, uthing: U) where ... {` even if it fits; if it fits that well, just put it directly in the generic parameters: `<T: ..., U: ...>`

**Comment by nrc** [comment]
> I'd assume, from your mention of visual indent, that when you say "each sub-clause on its own line", the first sub-clause shares the same line as the where?

yes

> One other observation: in the case where a function has no return value, I rather like this style:

I could live with this either way. I think it is fine to do this if there is no return type. Perhaps we should loo at some larger examples.

> Finally, I'd suggest explicitly saying "a where clause should always start a new line, even when it would fit on the previous line". I don't care for fn shortfunc<T, U>(thing: T, uthing: U) where ... { even if it fits; if it fits that well, just put it directly in the generic parameters: <T: ..., U: ...>

agreed

**Comment by joshtriplett** [comment]
@nrc
> I think it is fine to do this if there is no return type. Perhaps we should loo at some larger examples.

```rust
fn function_name<T>(
    param1: T,
    param2: Vec<T>,
) where T: Debug
{
    ...
}

fn function_name_2<I, T>(
    param1: I,
    param2: &mut Vec<T>,
) where I: IntoIterator<Item=T>
{
    ...
}
```

**Comment by withoutboats** [comment]
Over time I've come to prefer this style for where clauses:

```rust
item decl
where
    Type: Bound,
    Type: Bound,
{
    item body
}
```

I think this will also be better if we adopt the `where`-bound HRTB syntax:

```rust
where<T: Bound>,
    Type: Bound<T>,
    Type: Bound<T>,
```

It seems rather excessive to indent the where keyword and then indent again the clauses.

**Comment by nrc** [comment]
Fleshing out @withoutboats 's suggestion:

```
fn function_name<T>(
    param1: T,
    param2: Vec<T>,
)
where
    T: Debug
{
    let x = foo();
    x
}

fn function_name<T>(param1: T, param2: Vec<T>)
where
    T: Debug
{
    let x = foo();
    x
}

fn function_name_2<I, T>(
    param1: I,
    param2: &mut Vec<T>,
)
where
    I: IntoIterator<Item=T>,
    T: Debug,
{
    let x = foo();
    x
}
```

**Comment by nrc** [comment]
And fleshing out @joshtriplett 's examples to explore the contrast with functions with a return type:

```
fn function_name<T>(param1: T, param2: Vec<T>)
    where T: Debug
{
    let x = foo();
    x
}

fn function_name<T>(
    param1: T,
    param2: Vec<T>,
) where T: Debug
{
    let x = foo();
    x
}

fn function_name<T>(
    param1: T,
    param2: Vec<T>,
) -> ReturnType<T>
    where T: Debug
{
    let x = foo();
    x
}

fn function_name_2<I, T>(
    param1: I,
    param2: &mut Vec<T>,
) where I: IntoIterator<Item=T>,
        T: Debug,
{
    let x = foo();
    x
}
```

**Comment by nrc** [comment]
I think I don't like having `where` clauses left indented - to me it makes it harder to scan the left margin for items. I'm also not a fan of the newline before the first clause, seems like a waste of vertical space for no gain.

I think I'm still neutral on same-line `where` clauses where there is no return type - I like the more condensed look and I don't think it harms readability. OTOH, it is another minor inconsistency and perhaps makes it harder to scan for `where` clauses (though I'm not sure I've every done that).

**Comment by jplatte** [comment]
I really like having `where` at the same indentation level as the function it belongs to. I'm also slightly in favor of always having a newline before the first clause, not just for consistentcy, but because I don't like how cramped together this looks:

```rust
fn function_name<T>(
    param1: T,
    param2: Vec<T>,
) -> ReturnType<T>
    where T: Debug
{
    let x = foo();
    x
}

// alternatively, but still not really better IMHO
fn function_name<T>(
    param1: T,
    param2: Vec<T>,
) -> ReturnType<T>
where T: Debug
{
    let x = foo();
    x
}
```

**Comment by nrc** [comment]
We started discussing at the meeting but did not reach a conclusion. The options we were discussing were:

```
fn foo(
    x,
    y,
    z,
) -> Bloop 
    where
        Foo,
        Bar,
        Baz,
{
}

fn foo(
    x,
    y,
    z,
) -> Bloop 
where
    Foo,
    Bar,
    Baz,
{
}

fn foo(
    x,
    y,
    z,
) -> Bloop 
    where Foo,
          Bar,
          Baz
{
}
```

**Comment by nrc** [comment]
The first two are block-indented `where`, the last is visual indented, the first is indented from the left, the second is not.

My preference is for the third.

**Comment by nrc** [comment]
We did at least agree that we should allow the `where` clause on the same line as `)` if there is no return type.

**Comment by joshtriplett** [comment]
My preference is for the second: no extra level of indent for `where`, but an extra level of indent for its clauses (*if* they need breaking across multiple lines).

**Comment by nrc** [comment]
Oh, look at this (also from the meeting):

```
fn foo(x: i32, y: u32) where
    Foo,
    Bar,
{
}
```

I would prefer not to do this, keeping the where clause on its own line unless the args span multiple lines.

**Comment by nrc** [comment]
@joshtriplett the reason I dislike 2 is that for the single line where clause, you end up with a lot of left-indented code, which I find hard to scan:

```
fn foo(
    x,
    y,
    z,
) -> Bloop 
where Foo
{
    ...
}
```

I prefer having only the `fn` and `{` on the left - makes scanning the margin easer for me.

**Comment by joshtriplett** [comment]
@nrc I agree that putting the `where` on the same line as the `)` should only happen in the multi-line no-return-value case, where the `)` would otherwise appear on a line of its own.

Regarding the single-line case: hmmm, fair point.  I find myself inclined towards indenting `where` for the single-line case, but I dislike doing so in the multi-line case.  However, that seems wildly inconsistent.

**Comment by nrc** [comment]
I'd like to move forward here. The open questions I believe are:
* choose the overall formatting (visual vs block vs block indented, see https://github.com/rust-lang-nursery/fmt-rfcs/issues/38#issuecomment-274658173. My preference is visual)
* whether where clauses should have a trailing comma (always, never, only if spanning multiple lines. My preference is never)
* should we permit multiple clauses on one line (yes or no, I think no, I find this very hard to read)

Opinions?

**Comment by solson** [comment]
@nrc Why wouldn't we use trailing commas here? I would apply the same rule as elsewhere: include the trailing comma if it is followed by a newline.

**Comment by joshtriplett** [comment]
@solson ...unless the whole where clause fits on a single line.  I don't think we should include the trailing comma when not splitting the where clause across multiple lines.

@nrc I would advocate for:

- block-indent
- trailing comma only if spanning multiple lines
- Allow multiple clauses on one line, only break across lines if needed.

For that last one, consider simple cases like this:

```rust
fn f<T, U>(x: Vec<T>, y: U) -> usize
    where T: Ord, U: ?Sized
```

I don't think that where clause needs breaking across multiple lines.

**Comment by nrc** [comment]
@solson I'm having trouble explaining why we'd skip the commas here, the best I can do is similar to something Josh said higher up that it feels like the whole `where` clause is a kind of block, just without an actual brace, and with visual indent the invisible brace is on the same line:

```
where { foo }

where { foo,
        bar }
```

I think with block indent, the 'invisible brace' would be on the next line and we'd have a trailing comma.

More practically, if there is a single clause and it all fits on one line, then I think it looks a bit silly to have the trailing comma, and I want to be consistent about that:

```
where foo

where foo,
      bar

// c.f.,

where foo,

where foo,
      bar,
```

**Comment by nrc** [comment]
@solson and @joshtriplett 

To clarify, I don't strongly oppose block indent (I feel a bit bad about the wasted vertical space, but it is more consistent with other stuff) and if we go for block indent then I would support a trailing comma.

**Comment by nrc** [comment]
@joshtriplett I pretty strongly oppose having single line where clauses with multiple sub-clauses, I think they can be *very* hard to read, e.g.,

```
fn function_name_2<I, T>(
    param1: I,
    param2: &mut Vec<T>,
) where I: IntoIterator<Item=T>, for<'a> T: FooBar<'a>, IntoIterator::Item: FooBar<'static>
{
    ...
}
```

**Comment by nrc** [comment]
c.f. with newlines:

```
fn function_name_2<I, T>(
    param1: I,
    param2: &mut Vec<T>,
) where
    I: IntoIterator<Item=T>,
    for<'a> T: FooBar<'a>,
    IntoIterator::Item: FooBar<'static>,
{
    ...
}
```

**Comment by joshtriplett** [comment]
@nrc Would this be a good case for "short"/"simple"? :)

**Comment by nrc** [comment]
Do you have examples of short/simple where clauses where you'd keep them on one line? For me, they always seem a bit punctuation heavy even in the simplest cases.

**Comment by joshtriplett** [comment]
@nrc I included an example in https://github.com/rust-lang-nursery/fmt-rfcs/issues/38#issuecomment-278741804

They seem no more punctuation-heavy than function declaration parameter lists.

**Comment by nrc** [comment]
Hmm, true, I would write

```
fn f<T: Ord, U: ?Sized>(x: Vec<T>, y: U) -> usize
```

but also 

```
fn f<T, U>(x: Vec<T>, y: U) -> usize
    where T: Ord,
          U: ?Sized
```

Which I guess is inconsistent.

On the other hand, I'd argue that if a where clause is simple enough to be on one line then it should be bounds on the generics and not a where clause at all. IOW, the fact that a where clause exists implies that it passes (or fails, depending on your point of view) the short/simple test.

**Comment by joshtriplett** [comment]
Summarizing the consensus from the style meeting:

```rust
// With arguments fitting on one line:
fn function<T, U>(args)
where
    T: Bound,
    U: AnotherBound,
{
    body
}

// With arguments split onto their own lines, and no return type:
fn foo<T, U>(
    args,
) where
    T: Bound,
    U: AnotherBound,
{
    body
}

// With arguments split onto their own lines, and a return type:
fn foo<T, U>(
    args
) -> ReturnType
where
    T: Bound,
    U: AnotherBound,
{
    body
}
```

Note that we never put a bound (even a single bound) on the sme line as `where`, or multiple bounds on the same line as each other.

If bounds seem short and simple enough that this seems to add too much vertical space, consider putting them directly in the generics `<>` rather than in a separate where clause.

**Comment by solson** [comment]
In addition, the style @joshtriplett summarized avoids the potential future problem @withoutboats brought up regarding `where<T: Bound>` and visual alignment.

**Comment by nrc** [comment]
A note, I think if the whole item ends with a `;` we should skip the trailing comma. E.g.,

```
fn foo<T, U>(
    args
) -> ReturnType
where
    T: Bound,
    U: AnotherBound;
```

not 

```
where
    T: Bound,
    U: AnotherBound,;
```

An alternative would be to put the `;` on its own line, but I think we never otherwise do this.

**Comment by joshtriplett** [comment]
@nrc Agreed, that looks like the right way to format standalone declarations without a function body.

**Comment by nrc** [comment]
This is now implemented in Rustfmt, you'll need `where_style = "Rfc"` in your rustfmt.toml (or see the rfc-rustfmt.toml file in the root of the rustfmt repo).

You can see what this looks like in practice at https://github.com/rust-lang-nursery/rustfmt/blob/master/tests/target/where-clause-rfc.rs

**Comment by dotdash** [comment]
Like @stebalien says here: https://users.rust-lang.org/t/new-style-guidelines/10037/16 I quite dislike the fact that `where` while being a child of the `fn`/`impl` still has the same indentation as the `fn`/`impl` itself instead of being indented further.

**Comment by Stebalien** [comment]
Specifically, I'm suggesting indenting `where` by two spaces:

```rust
pub fn itemize_list<'a, T, I, F1, F2, F3>(
    codemap: &'a CodeMap,
    inner: I,
    terminator: &'a str,
    get_lo: F1,
    get_hi: F2,
    get_item_string: F3,
    prev_span_end: BytePos,
    next_span_start: BytePos,
) -> ListItems<'a, I, F1, F2, F3>
  where
    I: Iterator<Item = T>,
    F1: Fn(&T) -> BytePos,
    F2: Fn(&T) -> BytePos,
    F3: Fn(&T) -> Option<String>,
{
    ListItems {
        codemap: codemap,
        inner: inner.peekable(),
        get_lo: get_lo,
        get_hi: get_hi,
        get_item_string: get_item_string,
        prev_span_end: prev_span_end,
        next_span_start: next_span_start,
        terminator: terminator,
    }
}
```

One clear downside is that it introduces a "half" indent. One could indent `where` by four spaces (and then the constraints by another four) but that starts to get a bit deep:

```rust
pub fn itemize_list<'a, T, I, F1, F2, F3>(
    codemap: &'a CodeMap,
    inner: I,
    terminator: &'a str,
    get_lo: F1,
    get_hi: F2,
    get_item_string: F3,
    prev_span_end: BytePos,
    next_span_start: BytePos,
) -> ListItems<'a, I, F1, F2, F3>
    where
        I: Iterator<Item = T>,
        F1: Fn(&T) -> BytePos,
        F2: Fn(&T) -> BytePos,
        F3: Fn(&T) -> Option<String>,
{
    ListItems {
        codemap: codemap,
        inner: inner.peekable(),
        get_lo: get_lo,
        get_hi: get_hi,
        get_item_string: get_item_string,
        prev_span_end: prev_span_end,
        next_span_start: next_span_start,
        terminator: terminator,
    }
}
```

**Comment by nrc** [comment]
Copying @joshtriplett's summary of some of the decisions from meetings:

> During the style team meeting, we experimented with several different styles, both block and indent. We seriously considered visual style here, despite using block indent almost everywhere else, specifically because the keyword where has a fixed width so it wouldn't result in as many ripple effects across lines when editing.

> One of the bigger motivating factors to stick with block indent: without it, the where clause has a tendency to blend into the function signature and return type. When you have a complex type in the last argument, then a complex type in the return type, and then a complex type in the where clause, the three run together and make it hard to scan through as quickly. Each time we considered a rule with a visual indent, and applied it to a sample like that, we ended up finding the result harder to read. With the where keyword on a line by itself (or on the same line as the close paren when you have no return type), the where clause ends up more visually offset and easier to read at a glance.

**Comment by nrc** [comment]
I would be amenable to using a full (but not half) indent for `where`. I see nothing bad about it. IIRC we thought it was not important to indent and left alignment would be fine. But, from the users thread there seems to be some objection to this, so using an indent seems fine to me.

I prefer to avoid half indents. I think they are kinda ugly, it makes working with custom indent widths more difficult, and it can confuse text editors (which try to infer the indent width).

**Comment by bradleygibson** [comment]
I too am in support of the (full) indent for `where`.  Since the where is a specification of types begun by the fn, as a continuation, I would expect to see it indented, consistent with any other continuation--these are usually (always?) indented.

This greatly aids visual scanning of real code (not the trivial examples with one function consisting of single-character parameters names that are sometimes being shown here--those are too 'clean'--almost any format would look OK in that scenario).

Just my 2 cents,
-Brad

**Comment by jplatte** [comment]
So if we had a full indent for `where` for the cases where it goes onto its own line, what happens in the case of a multiline parameter list and no return type? Does `where` still go on the same line as the `)`, and if so, how far indented are the clauses?

```rust
// 1. where on the same line, clauses indented once
fn complicated_function<'a, T, U>(
    first_long_argument_name: SomeType<T>,
    second_long_argument_name: AnotherType<U>,
) where
    T: SomeTrait<AssocType = u32>,
    U: Iterator<Item = &'a T>,
{
    …
}

// 2. where on the same line, clauses indented twice
fn complicated_function<'a, T, U>(
    first_long_argument_name: SomeType<T>,
    second_long_argument_name: AnotherType<U>,
) where
        T: SomeTrait<AssocType = u32>,
        U: Iterator<Item = &'a T>,
{
    …
}

// 3. where on new line, same as when there is a return type
fn complicated_function<'a, T, U>(
    first_long_argument_name: SomeType<T>,
    second_long_argument_name: AnotherType<U>,
)
    where
        T: SomeTrait<AssocType = u32>,
        U: Iterator<Item = &'a T>,
{
    …
}
```

**Comment by rpjohnst** [comment]
Was there a reason to put `where` on its own line only when the function has a return type? I like the single-intended clauses with a newline before the opening brace, but it seems inconsistent to have the opening delimiter (the `where` keyword) on its own line when all other opening delimiters are on the previous line.

For example, I would prefer this:
```rust
fn complicated_function<'a, T, U>(
    first_long_argument_name: SomeType<T>,
    second_long_argument_name: AnotherType<U>,
) -> ReturnType<T> where
    T: SomeTrait<AssocType = u32>,
    U: Iterator<Item = &'a T>,
{
    …
}
```

**Comment by jplatte** [comment]
> when all other opening delimiters are on the previous line

That's not actually the case though. If you were to enforce that, the curly brace that starts the function body would have to be on the line of the last where clause in your example. And then it would be very hard to see where the clauses end and the function body begins.

**Comment by Stebalien** [comment]
> Was there a reason to put where on its own line only when the function has a return type?

Yes, it makes it harder to distinguish between the return type and the where clause. All variants of the current proposal have short lines without any types separating out the where constraints block.

```
...Type // arg
...Type // arg
...Type // arg
(...Type) // return?
      -- no types --
...Type // constraint
...Type // constraint
     -- no types --
...code...

**Comment by rpjohnst** [comment]
> That's not actually the case though.

I consider the opening curly brace also as the closing delimiter for the where clause, since there's no `endwhere`. If there were, I imagine it would go on the next line along with the opening curly brace.

> All variants of the current proposal have short lines without any types separating out the where constraints block.

What's the difference between that and what the current proposal already does with short argument lists? Should the `where` also go on its own line here?

```rust
fn short<T, U>(x: T, y: U) -> i32 where
    T: TraitWithLongName,
    U: UraitWithLongName,
{
    ...
}
```

**Comment by Stebalien** [comment]
> Should the where also go on its own line here?

According to the current proposal, [yes](https://github.com/rust-lang-nursery/fmt-rfcs/blob/master/example/lists-elided.rs#L330-L336).

**Comment by joshtriplett** [comment]
@nrc 
> I would be amenable to using a full (but not half) indent for where. I see nothing bad about it. IIRC we thought it was not important to indent and left alignment would be fine. But, from the users thread there seems to be some objection to this, so using an indent seems fine to me.

We did actually have a couple of reasons for not doing so.  The biggest one: it would mean that either functions with return types and functions without return types would indent the where clauses differently (meaning that adding or removing a return type would require reindenting the entire where clause, an odd inconsistency), or that where clauses for functions with no return type would have a double-indent and an odd lack of overlap (because `)where` is only 7 characters long and a double indent has 8 spaces):

```rust
) where
        T: SomeComplicatedTrait<With, Parameters>,
```

**Comment by emk** [comment]
I'm really happy with the fmt RFCs overall! The decisions are not all always what I would have chosen, but they seem reasonable and I'm sure I'll get used to them.

My biggest concern is definitely the `where` formatting, and the amount of vertical space it uses. For example:

```rust
where
      S1: AsRef<str>,
      S2: Into<String>,
      P: AsRef<Path>,
```

Especially when the constraints are short, I would personally write this as:

```rust
where S1: AsRef<str>, S2: Into<String>, P: AsRef<Path>
```

This is by analogy to short parameter lists:

```rust
// We write:
fn foo(s1: AsRef<str>, s2: Into<String>, p: AsRef<Path>)

// And not:
fn foo(
    s1: AsRef<str>,
    s2: Into<String>,
    p: AsRef<Path>
)
```

As a compromise, this is reasonable:

```rust
where S1: AsRef<str>,
        S2: Into<String>,
        P: AsRef<Path>,
```

...but using block formatting will use an extra line of vertical whitespace, and it only gains two horizontal spaces.

I'm relatively concerned about vertical space, because my eyes aren't getting any younger, and I can only comfortably fit 52 vertical lines of code in my editor on a good laptop. And that's already a slightly unpleasantly small font if I'm going to stare at it for hours.

And there's a real benefit to being able to see lots of code at once, and maintaining a "global overview" of what a program is doing.

**Comment by joshtriplett** [comment]
@emk I think the "short" case is the item we talked about longest.  I do actually favor the formatting you suggest (putting all three of those on one line).

The conclusion we ended up coming to was that if the constraints are that short or simple, you could put them directly in the generics, which *can* go on one line if they fit.  That works even better now that we use block indentation for arguments and don't try to align them after the open paren; a larger generics block won't push all the arguments off the right margin.  Inline simple constraints in the generics; use where clauses for longer or more complex constraints.

**Comment by bradleygibson** [comment]
> So if we had a full indent for where for the cases where it goes onto its own line, what happens in the case of a multiline parameter list and no return type?
IMHO, @jplatte When the where goes on its own line (as per the condition in your question), option #3.

**Comment by DorianListens** [comment]
I'd also be in favor of always putting where on its own line, with full indent. It's an important keyword, and it's important to separate where clauses from argument lists. I don't like the look of the where clause at the same indentation level as the function it acts upon, that seems like a very odd visual hierarchy mismatch.

**Comment by QuietMisdreavus** [comment]
I'm currently updating rustdoc to conform to the newer style guidelines and I ran into a fun wrinkle: Type aliases can also have where clauses. Even better, they go between the alias name and the expanded form.

When written inline, a type alias with a where clause looks like this:

```rust
type TestAlias<T> where T: Copy = Vec<T>;
```

With the proposed formatting, it seems like it should be somewhat like this:

```rust
type FormattedAlias<T>
where
    T: Copy
= Vec<T>;
```

Is this considered proper? That `= Vec<T>` on its own line looks really awkward, but without changing the grammar to move the where clause after the expanded form (is that the right term?), that seems to be the way to go. It doesn't help that type aliases don't check their type constraints, either.

**Comment by joshtriplett** [comment]
@QuietMisdreavus Interesting.  I think I'd ideally write it all on one line *without* a where clause (`type Something<T: Copy> = Vec<T>;`).  But in the case where you *do* want a where clause, that doesn't look like reasonable formatting to me.  I don't see a better place to put the `=`, though; having it start a line seems better than hanging it off the same line as the last where clause.  (Also, the where clauses should still have trailing commas when they appear on their own lines.)

**Comment by QuietMisdreavus** [comment]
Right, I didn't want it to hang off of the last condition, for the same reason as people talking about the opening brace earlier in the thread. I'll let it work like that for now. (And my mistake on the commas; I haven't absorbed that one into my personal style so I forgot them when sketching out the example. `>_>`)

**Comment by withoutboats** [comment]
Also should mention that the where clauses on type aliases don't *do* anything right now.

**Comment by nrc** [comment]
The only alternative I see is indenting all the following lines:

```
type FormattedAlias<T>
    where
        T: Copy,
    = Vec<T>;
```

I think @QuietMisdreavus's suggestion fits our rules better, but having everything on the left margin is annoying.

I agree that getting rid of the where clause is better than both.

**Comment by alteous** [comment]
I think the following style (adapted from @joshtriplett 's post / @withoutboats ' suggestion) is best:
```rust
fn foo<'a, T, U>(
    arg0: &'a Arg0,
    arg1: Arg1,
) -> ReturnType
where
    T: 'a + Bound,
    U: AnotherBound,
{
    body
}
```
#### Rationale
 - Quick to parse mentally since the `where` keyword is a visual barrier between the parameter declarations and the type bounds.
 - Indentation is consistent every time.
 - Suitable for long types.

There maybe a few extra newlines than absolutely necessary, but I think this is a very reasonable trade-off for the gains in readability.

Note that if there's only one type parameter then there's no need for the `where` clause in the first place:
```rust
fn foo<T: Bound>(arg0: Arg0) -> ReturnType {
    body
}
```

**Comment by joshtriplett** [comment]
We discussed this further in the style meeting, and https://github.com/rust-lang-nursery/fmt-rfcs/issues/38#issuecomment-284576257 is still an accurate summary of the style we've got consensus around.

Regarding type aliases (as raised by @QuietMisdreavus), I don't think we have a complete consensus on those.  We'd like to discuss type aliases and the possibility of single-line where clauses in #74.  I think we have consensus that at a minimum, there should be an option available for that; we don't, however, have consensus on that being the default.
